### PR TITLE
jpeg-xl: enable jpegli target

### LINKED ISSUE #3360: cjpegli not included in Homebrew ?
**Describe the bug**
Was reading about the fast decode times of jpegli on Cloudinary and wanted to test, if it has a positive impact on browser rendering times in comparison to mozjpg.
Installed Jpeg-XL via Homebrew without any errors and ran the cjpegli command.

zsh: command not found: cjpegli

**To Reproduce**
- Have Homebrew installed
- Open Terminal 
- Run "brew install jpeg-xl"
- Run "cjpegli input.jpg output.jpg"

**Expected behavior**
After a successful install through homebrew, I was expecting the cjpegli command to be available.

**Screenshots**
Screenshot 2024-03-04 at 13 57 41

**Environment**
 - OS: MacOs Monterey 12.7.3
 - CPU type: Intel Core i7 
 - cjxl/djxl version string: cjxl v0.10.1 0.10.1 [AVX2,SSE4,SSE2]

**Additional context**
Not sure if it has anything to do with it, but on line 62 the brew formula has the following option 
"-DJPEGXL_ENABLE_JPEGLI=OFF",

https://github.com/Homebrew/homebrew-core/blob/6b90bd87b55c14dd8cc04829be39f41e59b90b78/Formula/j/jpeg-xl.rb

Other people experiencing the same problem:
https://news.ycombinator.com/item?id=39568890
https://github.com/orgs/Homebrew/discussions/5074

### PR DESCRIPTION
**Pull request by Jerry1144**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

Found requests for jpegli target. Copied several headers from `jpeg-turbo` to fix build. Maybe upstream would flesh out the search logic in jpegli.cmake in a future version.

Fixes https://github.com/Homebrew/homebrew-core/pull/161210
Fixes https://github.com/libjxl/libjxl/issues/3360
Circumvents https://github.com/libjxl/libjxl/issues/3440

~~Disclaimer: I don't know what jpegli does, nor do I understand why there's added .in suffix of `jconfig.h` from `jpeg-turbo/include`. I can only confirm that these tools appear to work, and pass tests. Maybe we should test decode of jpegli instead of compression? Were these routines also provided in libs and should be tested by compiling a minimal binary like the other test?~~

### CODE REVIEW

File: Formula/j/jpeg-xl.rb [hunk 1]
```diff
@@ -54,12 +55,17 @@ class JpegXl < Formula
   def install
     ENV.append_path "XML_CATALOG_FILES", HOMEBREW_PREFIX/"etc/xml/catalog"
     resources.each { |r| r.stage buildpath/"third_party"/r.name }
+    # jpegli.cmake looks for these `jpeg-turbo` headers at hardcoded path
+    cp Formula["jpeg-turbo"].include/"jconfig.h",
+       buildpath/"third_party/libjpeg-turbo/jconfig.h.in"
+    cp [Formula["jpeg-turbo"].include/"jpeglib.h", Formula["jpeg-turbo"].include/"jmorecfg.h"],
```
Comment (Jerry1144) [inline Formula/j/jpeg-xl.rb:61]: I was tempted to write `Formula["jpeg-turbo"].include/["jpeglib.h", "jmorecfg.h"]` but that doesn't work (No implicit conversion from Array to String). Ruby actually handled `cp Array[src] dest` the intuitive way, nice.

File: Formula/j/jpeg-xl.rb [hunk 2]
```diff
@@ -54,12 +55,17 @@ class JpegXl < Formula
   def install
     ENV.append_path "XML_CATALOG_FILES", HOMEBREW_PREFIX/"etc/xml/catalog"
     resources.each { |r| r.stage buildpath/"third_party"/r.name }
+    # jpegli.cmake looks for these `jpeg-turbo` headers at hardcoded path
```
Comment (carlocab) [inline Formula/j/jpeg-xl.rb:58]: Can we fix this upstream instead?

Comment (Jerry1144) [reply Formula/j/jpeg-xl.rb:58]: I made a mistake and assumed `jconfig.h.in` == `jconfig.h`, when it is not - the [.in file] has several constants awaiting to be filled out during configure stage: `#define JPEG_LIB_VERSION  @JPEG_LIB_VERSION@` while the delivered config header from `jpeg-turbo` reflects that particular instance: `#define JPEG_LIB_VERSION  80`

[.in file]:https://github.com/libjpeg-turbo/libjpeg-turbo/blob/2dfe6c0fe9e18671105e94f7cbf044d4a1d157e6/jconfig.h.in#L4

`jpegli` tries to provide an ABI/API compatible library to libjpeg, and seems to just reuse code from jpeg-turbo (who also provides an ABI/API compatible libjpeg), not 'including' that. These headers were already downloaded during a source build of libjxl, so I think it's totally reasonable for upstream to assume these exist?

I think the best way would be a resource block, downloading that exact `jconfig.h.in` to that exact place. The other two headers are untouched by building process and can copied from `jpeg-turbo` or as a resource too.

**Comment by Jerry1144** [comment]
1. currently jpegli is hardcoded not to build a libjpeg dylib on Windows or Mac - [lib/jpegli.cmake#L110](https://github.com/libjxl/libjxl/blob/f78c90b9992f2c9adcb2da80b1a012e21e55c3c9/lib/jpegli.cmake#L110): `if (JPEGXL_ENABLE_JPEGLI_LIBJPEG AND NOT APPLE AND NOT WIN32 AND NOT EMSCRIPTEN)`
2. there is a static `libjpegli-static.a` artifact, but only the tools `cjpegli` and `djpegli` using it are installed by CMake. So either:
  - 2.1. keep things as-is. The three header files copied from `jpeg-turbo` will not be relevant. Copying them even if they should have been changed during configure, is acceptable to me.
  - 2.2 `lib.install "build/lib/libjpegli-static.a"` so anyone seriously interested in jpegli on macOS can also grab them. Very broken - need to cherry-pick headers (these under `include/jpegli` are for the libjpeg substitute, not the underlying library), manually construct a pkgconfig file (this step is under the 'build dynamic library' `if` so not done on Apple platform)

I took path 2.1, but still added this resource requirement because these headers from the `jpeg-turbo` formula cannot be used directly, but hacking them are too verbose. Good for benchmark on macOS I believe.

**Comment by Jerry1144** [comment]
OK, I guess there's no way to provide jpegli's `libjpeg` easily. If I enable JPEGXL_INSTALL_JPEGLI_LIBJPEG it conflicts with `jpeg-turbo` which is the default libjpeg version now. If I do not, no libraries from jpegli are included so users can only use `cjpegli` and `djpegli` tools, even from Linuxbrew. Making the entirety of `jpeg-xl` keg-only, then, will prevent access to the CLI tools.

Linuxbrew users probably need to unlink `jpeg-turbo`, manually change that line to `ON`, and rebuild, to get `jpeg-xl` install jpegli libraries and libjpeg. Or just source build with `--keep-tmp`, and scavenge for the libs under buildpath/build/jpegli

**Comment by Bustycat** [comment]
Google Open Source has announced Jpegli.
https://opensource.googleblog.com/2024/04/introducing-jpegli-new-jpeg-coding-library.html

**Comment by Jerry1144** [comment]
> Google Open Source has announced Jpegli. https://opensource.googleblog.com/2024/04/introducing-jpegli-new-jpeg-coding-library.html

Which still is a sub-library under `jpeg-xl`, and only builds its shared library for Linux, unless more updates come to that sub-directory in the following months.
***
The dependents check are timing out for this formula. I see no way how the libjxl part may change once the libjpegli part is compiled at build time, so maybe just skip these checks?

File: Formula/j/jpeg-xl.rb [hunk 3]
```diff
@@ -50,16 +51,25 @@ class JpegXl < Formula
     url "https://github.com/webmproject/sjpeg.git",
         revision: "e5ab13008bb214deb66d5f3e17ca2f8dbff150bf"
   end
+  resource "jpeg-turbo" do
```
Comment (SMillerDev) [inline Formula/j/jpeg-xl.rb:54]: There is already a formula for this

Comment (Jerry1144) [reply Formula/j/jpeg-xl.rb:54]: It uses *source* from jpeg-turbo, not its delivered header. Specifically a header.in file that gets filled at compile-time.

**Comment by CoolOppo** [comment]
Oh no don't close

**Comment by SuperBo** [comment]
Can we reopen this pull request?

**Comment by Bustycat** [comment]
Now Jpegli gets its own repo https://github.com/google/jpegli

**Comment by Jerry1144** [comment]
> Now Jpegli gets its own repo https://github.com/google/jpegli

There's no way to build jpegli only, it seems. There's also no versioning either. Hmm, is building a libjxl alongside and outright discarding it acceptable by Homebrew standards🤣? If so I could trivially modify my formula into `jpegxl.rb` and open a PR for it.

**Comment by andrews05** [comment]
@Jerry1144 Did you get any further with this? Even just including the cjpegli and djpegli tools without libjpeg would still be helpful.

**Pull request closed without merging**